File: papers/arcs_complete_outside_conic/check_q11_structure.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <tuple>
#include <vector>

// Independent prime-field replay of check_q11_structure.py.  This implementation
// shares no project code or generated data beyond the six published coordinates.

namespace {
constexpr int q = 11;
using P = std::array<int, 3>;
using G = std::array<int, 4>;
using M3 = std::array<std::array<int, 3>, 3>;

constexpr std::array<P, 6> arc{{
    P{1, 10, 0}, P{1, 9, 1}, P{1, 4, 7},
    P{1, 8, 5}, P{0, 1, 4}, P{1, 1, 7},
}};

int mod(int x) {
  x %= q;
  return x < 0 ? x + q : x;
}

int inv(int x) {
  x = mod(x);
  for (int y = 1; y < q; ++y) {
    if (x * y % q == 1) return y;
  }
  assert(false);
  return 0;
}

int powmod(int x, int n) {
  int out = 1;
  for (; n; n >>= 1, x = mod(x * x)) if (n & 1) out = mod(out * x);
  return out;
}

template <std::size_t n>
std::array<int, n> normalize(std::array<int, n> x) {
  auto it = std::find_if(x.begin(), x.end(), [](int y) { return mod(y) != 0; });
  assert(it != x.end());
  const int z = inv(*it);
  for (int& y : x) y = mod(z * y);
  return x;
}

int det(P a, P b, P c) {
  return mod(a[0] * (b[1] * c[2] - b[2] * c[1])
      - a[1] * (b[0] * c[2] - b[2] * c[0])
      + a[2] * (b[0] * c[1] - b[1] * c[0]));
}

P add(P a, P b) {
  for (int i = 0; i < 3; ++i) a[i] = mod(a[i] + b[i]);
  return a;
}

P scale(int c, P a) {
  for (int& x : a) x = mod(c * x);
  return a;
}

int encode(P a) { return a[0] * q * q + a[1] * q + a[2]; }

std::vector<P> points() {
  std::vector<P> out;
  for (int y = 0; y < q; ++y)
    for (int z = 0; z < q; ++z) out.push_back({1, y, z});
  for (int z = 0; z < q; ++z) out.push_back({0, 1, z});
  out.push_back({0, 0, 1});
  return out;
}

bool on_conic(P p) { return mod(p[0] * p[2]) == mod(p[1] * p[1]); }

template <typename Container>
bool is_arc(const Container& a) {
  std::set<P> distinct(a.begin(), a.end());
  if (distinct.size() != a.size()) return false;
  for (std::size_t i = 0; i < a.size(); ++i)
    for (std::size_t j = i + 1; j < a.size(); ++j)
      for (std::size_t k = j + 1; k < a.size(); ++k)
        if (det(a[i], a[j], a[k]) == 0) return false;
  return true;
}

template <typename Container>
int secant_index(P p, const Container& a) {
  int result = 0;
  for (std::size_t i = 0; i < a.size(); ++i)
    for (std::size_t j = i + 1; j < a.size(); ++j)
      result += det(a[i], a[j], p) == 0;
  return result;
}

template <typename Container>
std::vector<P> extensions(const Container& a, const std::vector<P>& plane) {
  std::set<P> aset(a.begin(), a.end());
  std::vector<P> out;
  for (P p : plane) {
    if (aset.count(p)) continue;
    bool good = true;
    for (std::size_t i = 0; i < a.size() && good; ++i)
      for (std::size_t j = i + 1; j < a.size(); ++j)
        if (det(a[i], a[j], p) == 0) good = false;
    if (good) out.push_back(p);
  }
  return out;
}

int quadratic_rank(const std::array<P, 6>& a) {
  std::array<std::array<int, 6>, 6> m{};
  for (int i = 0; i < 6; ++i) {
    const auto [x, y, z] = a[i];
    m[i] = {mod(x * x), mod(x * y), mod(x * z), mod(y * y), mod(y * z), mod(z * z)};
  }
  int r = 0;
  for (int c = 0; c < 6; ++c) {
    int pivot = r;
    while (pivot < 6 && m[pivot][c] == 0) ++pivot;
    if (pivot == 6) continue;
    std::swap(m[r], m[pivot]);
    int z = inv(m[r][c]);
    for (int j = c; j < 6; ++j) m[r][j] = mod(z * m[r][j]);
    for (int i = 0; i < 6; ++i) if (i != r && m[i][c]) {
      z = m[i][c];
      for (int j = c; j < 6; ++j) m[i][j] = mod(m[i][j] - z * m[r][j]);
    }
    ++r;
  }
  return r;
}

struct SyndromeSummary {
  std::map<int, int> distances;
  std::map<int, int> leaders2;
  auto operator<=>(const SyndromeSummary&) const = default;
};

template <typename Container>
SyndromeSummary syndrome_summary(const Container& a) {
  std::array<int, q * q * q> distance;
  std::array<int, q * q * q> count2{};
  distance.fill(4);
  distance[0] = 0;
  for (P x : a) for (int c = 1; c < q; ++c)
    distance[encode(scale(c, x))] = std::min(distance[encode(scale(c, x))], 1);
  for (std::size_t i = 0; i < a.size(); ++i)
    for (std::size_t j = i + 1; j < a.size(); ++j)
      for (int c = 1; c < q; ++c) for (int d = 1; d < q; ++d) {
        P s = add(scale(c, a[i]), scale(d, a[j]));
        ++count2[encode(s)];
        distance[encode(s)] = std::min(distance[encode(s)], 2);
      }
  for (std::size_t i = 0; i < a.size(); ++i)
    for (std::size_t j = i + 1; j < a.size(); ++j)
      for (std::size_t k = j + 1; k < a.size(); ++k)
        for (int c = 1; c < q; ++c) for (int d = 1; d < q; ++d)
          for (int e = 1; e < q; ++e) {
            P s = add(add(scale(c, a[i]), scale(d, a[j])), scale(e, a[k]));
            distance[encode(s)] = std::min(distance[encode(s)], 3);
          }
  SyndromeSummary out;
  for (int s = 0; s < q * q * q; ++s) {
    assert(distance[s] <= 3);
    ++out.distances[distance[s]];
    if (distance[s] == 2) ++out.leaders2[count2[s]];
  }
  return out;
}

G gmul(G g, G h) {
  const auto [a, b, c, d] = g;
  const auto [e, f, k, l] = h;
  return normalize(G{a * e + b * k, a * f + b * l, c * e + d * k, c * f + d * l});
}

M3 sym2(G g) {
  const auto [a, b, c, d] = g;
  return {{{mod(a * a), mod(2 * a * b), mod(b * b)},
           {mod(a * c), mod(a * d + b * c), mod(b * d)},
           {mod(c * c), mod(2 * c * d), mod(d * d)}}};
}

P apply(M3 m, P p) {
  P out{};
  for (int i = 0; i < 3; ++i)
    for (int j = 0; j < 3; ++j) out[i] = mod(out[i] + m[i][j] * p[j]);
  return normalize(out);
}

std::vector<G> pgl2() {
  std::set<G> unique;
  for (int a = 0; a < q; ++a) for (int b = 0; b < q; ++b)
    for (int c = 0; c < q; ++c) for (int d = 0; d < q; ++d)
      if (mod(a * d - b * c)) unique.insert(normalize(G{a, b, c, d}));
  return {unique.begin(), unique.end()};
}

int order(G g) {
  const G one{1, 0, 0, 1};
  G x = one;
  for (int n = 1; n <= 120; ++n) {
    x = gmul(x, g);
    if (x == one) return n;
  }
  assert(false);
  return 0;
}

std::set<G> generated(G x, G y) {
  const G one{1, 0, 0, 1};
  std::set<G> seen{one};
  std::queue<G> todo;
  todo.push(one);
  while (!todo.empty()) {
    G z = todo.front();
    todo.pop();
    for (G g : {x, y}) {
      G w = gmul(z, g);
      if (seen.insert(w).second) todo.push(w);
    }
  }
  return seen;
}

template <typename Container>
std::vector<G> stabilizer(const Container& a, const std::vector<G>& pgl) {
  std::set<P> target(a.begin(), a.end());
  std::vector<G> out;
  for (G g : pgl) {
    std::set<P> image;
    for (P p : a) image.insert(apply(sym2(g), p));
    if (image == target) out.push_back(g);
  }
  return out;
}

std::vector<int> orbit_sizes(const std::vector<G>& group, const std::vector<P>& plane) {
  std::set<P> remaining(plane.begin(), plane.end());
  std::vector<int> sizes;
  while (!remaining.empty()) {
    P seed = *remaining.begin();
    std::set<P> orbit;
    for (G g : group) orbit.insert(apply(sym2(g), seed));
    sizes.push_back(static_cast<int>(orbit.size()));
    for (P p : orbit) remaining.erase(p);
  }
  std::sort(sizes.begin(), sizes.end());
  return sizes;
}

std::string map_text(const std::map<int, int>& xs) {
  std::string out;
  for (auto [k, v] : xs) {
    if (!out.empty()) out += ',';
    out += std::to_string(k) + ':' + std::to_string(v);
  }
  return out;
}
}  // namespace

int main() {
  const std::vector<P> plane = points();
  assert(plane.size() == 133);
  assert(is_arc(arc));
  std::vector<P> conic;
  for (P p : plane) if (on_conic(p)) conic.push_back(p);
  assert(conic.size() == 12);
  for (P p : arc) assert(!on_conic(p));

  std::map<int, int> required_indices, conic_indices;
  std::set<P> aset(arc.begin(), arc.end()), cset(conic.begin(), conic.end());
  for (P p : plane) {
    if (aset.count(p)) continue;
    if (cset.count(p)) ++conic_indices[secant_index(p, arc)];
    else ++required_indices[secant_index(p, arc)];
  }
  assert((required_indices == std::map<int, int>{{1, 90}, {2, 15}, {3, 10}}));
  assert((conic_indices == std::map<int, int>{{0, 12}}));
  assert(quadratic_rank(arc) == 6);

  const std::vector<P> ext = extensions(arc, plane);
  assert(std::set<P>(ext.begin(), ext.end()) == cset);
  const SyndromeSummary syndromes = syndrome_summary(arc);
  assert((syndromes.distances == std::map<int, int>{{0, 1}, {1, 60}, {2, 1150}, {3, 120}}));
  assert((syndromes.leaders2 == std::map<int, int>{{1, 900}, {2, 150}, {3, 100}}));

  int edges = 0;
  std::array<int, 12> degrees{};
  std::array<int, 6> color_edges{};
  std::array<std::set<int>, 6> color_vertices;
  std::set<std::pair<int, int>> edge_set;
  for (int i = 0; i < 12; ++i) for (int j = i + 1; j < 12; ++j) {
    std::vector<int> witnesses;
    for (int a = 0; a < 6; ++a) if (det(ext[i], ext[j], arc[a]) == 0) witnesses.push_back(a);
    if (!witnesses.empty()) {
      assert(witnesses.size() == 1);
      ++edges;
      ++degrees[i]; ++degrees[j];
      ++color_edges[witnesses[0]];
      color_vertices[witnesses[0]].insert(i);
      color_vertices[witnesses[0]].insert(j);
      edge_set.insert({i, j});
    }
  }
  assert(edges == 30);
  for (int d : degrees) assert(d == 5);
  for (int a = 0; a < 6; ++a) {
    assert(color_edges[a] == 5);
    assert(color_vertices[a].size() == 10);
    std::set<int> tangents;
    for (int i = 0; i < 12; ++i) {
      int intersections = 0;
      for (int j = 0; j < 12; ++j) intersections += det(arc[a], ext[i], ext[j]) == 0;
      if (intersections == 1) tangents.insert(i);
    }
    std::set<int> missing;
    for (int i = 0; i < 12; ++i) if (!color_vertices[a].count(i)) missing.insert(i);
    assert(missing == tangents && tangents.size() == 2);
  }

  const std::vector<G> pgl = pgl2();
  assert(pgl.size() == 1320);
  const std::vector<G> stab = stabilizer(arc, pgl);
  assert(stab.size() == 60);
  std::map<int, int> orders;
  for (G g : stab) {
    ++orders[order(g)];
    const int determinant = mod(g[0] * g[3] - g[1] * g[2]);
    assert(powmod(determinant, 5) == 1);
  }
  assert((orders == std::map<int, int>{{1, 1}, {2, 15}, {3, 20}, {5, 24}}));
  assert((orbit_sizes(stab, plane) == std::vector<int>{6, 10, 12, 15, 30, 30, 30}));

  G gen2{}, gen3{};
  bool found_generators = false;
  for (G x : stab) if (order(x) == 2) {
    for (G y : stab) if (order(y) == 3 && order(gmul(x, y)) == 5) {
      if (generated(x, y).size() == 60) {
        gen2 = x; gen3 = y; found_generators = true; break;
      }
    }
    if (found_generators) break;
  }
  assert(found_generators);

  std::set<P> arc_orbit, conic_orbit;
  for (G g : stab) {
    arc_orbit.insert(apply(sym2(g), arc[0]));
    conic_orbit.insert(apply(sym2(g), conic[0]));
  }
  assert(arc_orbit.size() == 6 && conic_orbit.size() == 12);
  std::set<std::pair<P, P>> edge_orbit;
  const auto [ei, ej] = *edge_set.begin();
  for (G g : stab) {
    P x = apply(sym2(g), ext[ei]), y = apply(sym2(g), ext[ej]);
    if (y < x) std::swap(x, y);
    edge_orbit.insert({x, y});
  }
  assert(edge_orbit.size() == 30);

  const M3 transform{{{{1, 1, 0}}, {{0, 1, 1}}, {{1, 0, 1}}}};
  std::array<P, 6> transformed_arc{};
  std::vector<P> transformed_conic;
  for (int i = 0; i < 6; ++i) transformed_arc[5 - i] = apply(transform, arc[i]);
  for (P p : conic) transformed_conic.push_back(apply(transform, p));
  const auto transformed_ext = extensions(transformed_arc, plane);
  assert(std::set<P>(transformed_ext.begin(), transformed_ext.end())
      == std::set<P>(transformed_conic.begin(), transformed_conic.end()));
  assert(quadratic_rank(transformed_arc) == 6);
  assert(syndrome_summary(transformed_arc) == syndromes);

  std::array<P, 6> perturbed = arc;
  perturbed[0] = {1, 0, 8};
  assert(is_arc(perturbed));
  const auto perturbed_ext = extensions(perturbed, plane);
  const auto perturbed_stab = stabilizer(perturbed, pgl);
  assert(perturbed_ext.size() == 20 && perturbed_stab.size() == 2);

  G mutated = gen2;
  mutated[0] = mod(mutated[0] + 1);
  if (mod(mutated[0] * mutated[3] - mutated[1] * mutated[2]) == 0 ||
      std::find(stab.begin(), stab.end(), normalize(mutated)) != stab.end()) {
    mutated = gen2;
    mutated[1] = mod(mutated[1] + 1);
  }
  assert(mod(mutated[0] * mutated[3] - mutated[1] * mutated[2]));
  assert(std::find(stab.begin(), stab.end(), normalize(mutated)) == stab.end());

  std::cout << "points=133 conic=12 arc=6 quadratic_rank=6"
            << " required_indices=" << map_text(required_indices)
            << " syndrome_distances=" << map_text(syndromes.distances)
            << " distance2_leaders=" << map_text(syndromes.leaders2)
            << " extensions=12 edges=30 degrees=5x12 colors=5x6 tangents=2x6"
            << " pgl2=1320 stabilizer=60 orders=" << map_text(orders)
            << " orbits=6,10,12,15,30,30,30"
            << " generators=2,3 product=5 generated=60"
            << " transitive=arc6,conic12,edge30"
            << " coordinate_invariance=PASS"
            << " negative=extensions20,stabilizer2,mutated_generator_rejected"
            << '\n';
}

File: papers/arcs_complete_outside_conic/check_q11_structure.py
#!/usr/bin/env python3
"""Independent exhaustive checks for the exceptional q=11 witness.

This script deliberately does not import the manuscript verifier.  It rebuilds
PG(2,11), the standard conic, syndrome errors, and PGL(2,11) from elementary
prime-field arithmetic.  Its output is deterministic JSON suitable for hashing.
"""

from __future__ import annotations

from collections import Counter, deque
from itertools import combinations, product
import json

Q = 11
Point = tuple[int, int, int]
PGL2 = tuple[int, int, int, int]

ARC: tuple[Point, ...] = (
    (1, 10, 0),
    (1, 9, 1),
    (1, 4, 7),
    (1, 8, 5),
    (0, 1, 4),
    (1, 1, 7),
)


def inv(x: int) -> int:
    return pow(x % Q, -1, Q)


def normalize(xs: tuple[int, ...]) -> tuple[int, ...]:
    ys = tuple(x % Q for x in xs)
    pivot = next(x for x in ys if x)
    scale = inv(pivot)
    return tuple(scale * x % Q for x in ys)


def det(u: Point, v: Point, w: Point) -> int:
    return (
        u[0] * (v[1] * w[2] - v[2] * w[1])
        - u[1] * (v[0] * w[2] - v[2] * w[0])
        + u[2] * (v[0] * w[1] - v[1] * w[0])
    ) % Q


def add(*vs: Point) -> Point:
    return tuple(sum(v[i] for v in vs) % Q for i in range(3))  # type: ignore[return-value]


def scale(c: int, v: Point) -> Point:
    return tuple(c * x % Q for x in v)  # type: ignore[return-value]


def rank_mod(rows: list[tuple[int, ...]]) -> int:
    a = [list(x % Q for x in row) for row in rows]
    if not a:
        return 0
    r = 0
    for c in range(len(a[0])):
        pivot = next((i for i in range(r, len(a)) if a[i][c]), None)
        if pivot is None:
            continue
        a[r], a[pivot] = a[pivot], a[r]
        z = inv(a[r][c])
        a[r] = [z * x % Q for x in a[r]]
        for i in range(len(a)):
            if i != r and a[i][c]:
                z = a[i][c]
                a[i] = [(x - z * y) % Q for x, y in zip(a[i], a[r])]
        r += 1
        if r == len(a):
            break
    return r


def projective_points() -> tuple[Point, ...]:
    return tuple(
        [(1, y, z) for y in range(Q) for z in range(Q)]
        + [(0, 1, z) for z in range(Q)]
        + [(0, 0, 1)]
    )


POINTS = projective_points()
CONIC = tuple(p for p in POINTS if p[0] * p[2] % Q == p[1] * p[1] % Q)


def is_arc(arc: tuple[Point, ...]) -> bool:
    return len(set(arc)) == len(arc) and all(det(*triple) for triple in combinations(arc, 3))


def secant_index(p: Point, arc: tuple[Point, ...]) -> int:
    return sum(det(a, b, p) == 0 for a, b in combinations(arc, 2))


def quadratic_rank(arc: tuple[Point, ...]) -> int:
    return rank_mod([
        (x * x, x * y, x * z, y * y, y * z, z * z)
        for x, y, z in arc
    ])


def extension_points(arc: tuple[Point, ...]) -> tuple[Point, ...]:
    aset = set(arc)
    return tuple(
        p for p in POINTS
        if p not in aset and all(det(a, b, p) for a, b in combinations(arc, 2))
    )


def syndrome_summary(arc: tuple[Point, ...]) -> dict[str, object]:
    zero: Point = (0, 0, 0)
    by_weight: list[Counter[Point]] = [Counter() for _ in range(4)]
    by_weight[0][zero] = 1
    for i, a in enumerate(arc):
        del i
        for c in range(1, Q):
            by_weight[1][scale(c, a)] += 1
    for i, j in combinations(range(len(arc)), 2):
        for c, d in product(range(1, Q), repeat=2):
            by_weight[2][add(scale(c, arc[i]), scale(d, arc[j]))] += 1
    for i, j, k in combinations(range(len(arc)), 3):
        for c, d, e in product(range(1, Q), repeat=3):
            by_weight[3][add(scale(c, arc[i]), scale(d, arc[j]), scale(e, arc[k]))] += 1

    distance: dict[Point, int] = {}
    for weight, syndromes in enumerate(by_weight):
        for s in syndromes:
            distance.setdefault(s, weight)
    assert len(distance) == Q**3
    dcounts = Counter(distance.values())
    leaders2 = Counter(by_weight[2][s] for s, d in distance.items() if d == 2)
    return {
        "distance_counts": dict(sorted(dcounts.items())),
        "distance_two_leader_histogram": dict(sorted(leaders2.items())),
    }


def mat3_apply(m: tuple[tuple[int, int, int], ...], p: Point) -> Point:
    return normalize(tuple(sum(m[i][j] * p[j] for j in range(3)) % Q for i in range(3)))  # type: ignore[return-value]


def pgl2_mul(g: PGL2, h: PGL2) -> PGL2:
    a, b, c, d = g
    e, f, k, l = h
    return normalize((a * e + b * k, a * f + b * l, c * e + d * k, c * f + d * l))  # type: ignore[return-value]


def sym2(g: PGL2) -> tuple[tuple[int, int, int], ...]:
    a, b, c, d = g
    return (
        (a * a % Q, 2 * a * b % Q, b * b % Q),
        (a * c % Q, (a * d + b * c) % Q, b * d % Q),
        (c * c % Q, 2 * c * d % Q, d * d % Q),
    )


def pgl2_elements() -> tuple[PGL2, ...]:
    return tuple(sorted({
        normalize((a, b, c, d))  # type: ignore[arg-type]
        for a, b, c, d in product(range(Q), repeat=4)
        if (a * d - b * c) % Q
    }))


def pgl2_order(g: PGL2) -> int:
    identity: PGL2 = (1, 0, 0, 1)
    x = identity
    for n in range(1, 121):
        x = pgl2_mul(x, g)
        if x == identity:
            return n
    raise AssertionError("PGL2 element order exceeded group order bound")


def generated_subgroup(generators: tuple[PGL2, ...]) -> set[PGL2]:
    identity: PGL2 = (1, 0, 0, 1)
    seen = {identity}
    todo = deque([identity])
    while todo:
        x = todo.popleft()
        for g in generators:
            y = pgl2_mul(x, g)
            if y not in seen:
                seen.add(y)
                todo.append(y)
    return seen


def action_orbits(group: tuple[PGL2, ...], objects: tuple[Point, ...]) -> list[tuple[Point, ...]]:
    remaining = set(objects)
    result: list[tuple[Point, ...]] = []
    while remaining:
        seed = min(remaining)
        orbit = {mat3_apply(sym2(g), seed) for g in group}
        assert orbit <= set(objects)
        result.append(tuple(sorted(orbit)))
        remaining -= orbit
    return result


def group_summary(arc: tuple[Point, ...]) -> tuple[dict[str, object], tuple[PGL2, ...]]:
    pgl = pgl2_elements()
    aset = set(arc)
    stabilizer = tuple(g for g in pgl if {mat3_apply(sym2(g), p) for p in arc} == aset)
    orders = Counter(pgl2_order(g) for g in stabilizer)
    determinant_square = Counter(
        pow((g[0] * g[3] - g[1] * g[2]) % Q, (Q - 1) // 2, Q)
        for g in stabilizer
    )

    by_order: dict[int, list[PGL2]] = {}
    for g in stabilizer:
        by_order.setdefault(pgl2_order(g), []).append(g)
    generators: tuple[PGL2, PGL2] | None = None
    for x in by_order.get(2, []):
        for y in by_order.get(3, []):
            if pgl2_order(pgl2_mul(x, y)) == 5 and len(generated_subgroup((x, y))) == 60:
                generators = (x, y)
                break
        if generators is not None:
            break
    assert generators is not None

    orbit_rows = []
    for orbit in action_orbits(stabilizer, POINTS):
        in_arc = orbit[0] in aset
        on_conic = orbit[0] in set(CONIC)
        indices = sorted({secant_index(p, arc) for p in orbit}) if not in_arc else []
        orbit_rows.append({
            "size": len(orbit),
            "arc": in_arc,
            "conic": on_conic,
            "secant_indices": indices,
            "representative": orbit[0],
        })
    orbit_rows.sort(key=lambda row: (not row["arc"], not row["conic"], row["size"], row["representative"]))
    return ({
        "pgl2_size": len(pgl),
        "stabilizer_size": len(stabilizer),
        "element_orders": dict(sorted(orders.items())),
        "determinant_legendre": dict(sorted(determinant_square.items())),
        "generators_order_2_3": generators,
        "generator_product_order": pgl2_order(pgl2_mul(*generators)),
        "generated_size": len(generated_subgroup(generators)),
        "point_orbits": orbit_rows,
    }, stabilizer)


def chord_summary(arc: tuple[Point, ...], extensions: tuple[Point, ...]) -> dict[str, object]:
    edges: list[tuple[Point, Point]] = []
    colors: list[list[tuple[Point, Point]]] = [[] for _ in arc]
    for p, q in combinations(extensions, 2):
        witnesses = [i for i, a in enumerate(arc) if det(p, q, a) == 0]
        if witnesses:
            assert len(witnesses) == 1
            edge = (p, q)
            edges.append(edge)
            colors[witnesses[0]].append(edge)

    degrees = Counter()
    for p, q in edges:
        degrees[p] += 1
        degrees[q] += 1

    color_rows = []
    for i, matching in enumerate(colors):
        used = [p for edge in matching for p in edge]
        assert len(used) == len(set(used))
        missing = sorted(set(extensions) - set(used))
        tangent_contacts = sorted(
            p for p in extensions
            if sum(det(arc[i], p, q) == 0 for q in extensions) == 1
        )
        assert missing == tangent_contacts
        color_rows.append({
            "witness": i,
            "edges": len(matching),
            "covered_vertices": len(used),
            "missing_tangent_contacts": missing,
        })

    edge_set = {frozenset(edge) for edge in edges}
    return {
        "edges": len(edges),
        "degree_histogram": dict(sorted(Counter(degrees.values()).items())),
        "color_classes": color_rows,
        "partition_distinct": sum(len(x) for x in colors) == len(edge_set),
    }


def transformed_invariance(arc: tuple[Point, ...], conic: tuple[Point, ...], stabilizer: tuple[PGL2, ...]) -> bool:
    transform = ((1, 1, 0), (0, 1, 1), (1, 0, 1))
    transformed_arc = tuple(reversed(tuple(mat3_apply(transform, p) for p in arc)))
    transformed_conic = tuple(mat3_apply(transform, p) for p in conic)
    transformed_extensions = extension_points(transformed_arc)
    if set(transformed_extensions) != set(transformed_conic):
        return False
    if quadratic_rank(transformed_arc) != quadratic_rank(arc):
        return False
    if syndrome_summary(transformed_arc) != syndrome_summary(arc):
        return False
    # Directly transport every stabilizer permutation; conjugate matrices are unnecessary here.
    original_orbits = sorted(len(o) for o in action_orbits(stabilizer, POINTS))
    transformed_images = {
        tuple(mat3_apply(transform, mat3_apply(sym2(g), p)) for p in arc)
        for g in stabilizer
    }
    return len(transformed_images) == 60 and original_orbits == [6, 10, 12, 15, 30, 30, 30]


def negative_control(stabilizer: tuple[PGL2, ...], generators: tuple[PGL2, PGL2]) -> dict[str, object]:
    original_extensions = extension_points(ARC)
    perturbed: tuple[Point, ...] | None = None
    for replacement in POINTS:
        candidate = (replacement,) + ARC[1:]
        if replacement not in CONIC and is_arc(candidate):
            if set(extension_points(candidate)) != set(original_extensions):
                perturbed = candidate
                break
    assert perturbed is not None
    perturbed_stabilizer = tuple(
        g for g in pgl2_elements()
        if {mat3_apply(sym2(g), p) for p in perturbed} == set(perturbed)
    )

    g = generators[0]
    mutated: PGL2 | None = None
    for i in range(4):
        raw = list(g)
        raw[i] = (raw[i] + 1) % Q
        if (raw[0] * raw[3] - raw[1] * raw[2]) % Q:
            candidate = normalize(tuple(raw))  # type: ignore[arg-type]
            if candidate not in stabilizer:
                mutated = candidate  # type: ignore[assignment]
                break
    assert mutated is not None
    return {
        "perturbed_first_point": perturbed[0],
        "perturbed_extension_count": len(extension_points(perturbed)),
        "perturbed_stabilizer_size": len(perturbed_stabilizer),
        "mutated_generator_preserves_arc": mutated in stabilizer,
    }


def main() -> None:
    assert len(POINTS) == 133 and len(set(POINTS)) == 133
    assert len(CONIC) == 12
    assert is_arc(ARC) and set(ARC).isdisjoint(CONIC)

    off_arc = [p for p in POINTS if p not in set(ARC)]
    required = [p for p in off_arc if p not in set(CONIC)]
    required_indices = Counter(secant_index(p, ARC) for p in required)
    conic_indices = Counter(secant_index(p, ARC) for p in CONIC)
    extensions = extension_points(ARC)
    assert set(extensions) == set(CONIC)

    group, stabilizer = group_summary(ARC)
    chords = chord_summary(ARC, extensions)
    syndrome = syndrome_summary(ARC)
    generators = group["generators_order_2_3"]
    assert isinstance(generators, tuple)

    edge_set = {
        frozenset((p, q)) for p, q in combinations(extensions, 2)
        if any(det(p, q, a) == 0 for a in ARC)
    }
    edge_orbit = {
        frozenset((mat3_apply(sym2(g), p), mat3_apply(sym2(g), q)))
        for g in stabilizer for p, q in [tuple(next(iter(edge_set)))]
    }

    result = {
        "field": Q,
        "points": len(POINTS),
        "conic_points": len(CONIC),
        "arc_points": len(ARC),
        "quadratic_evaluation_rank": quadratic_rank(ARC),
        "required_secant_indices": dict(sorted(required_indices.items())),
        "conic_secant_indices": dict(sorted(conic_indices.items())),
        "single_extensions": len(extensions),
        "syndromes": syndrome,
        "chords": chords,
        "group": group,
        "action_transitivity": {
            "arc_orbit": len({mat3_apply(sym2(g), ARC[0]) for g in stabilizer}),
            "conic_orbit": len({mat3_apply(sym2(g), CONIC[0]) for g in stabilizer}),
            "edge_orbit": len(edge_orbit),
        },
        "coordinate_invariance": transformed_invariance(ARC, CONIC, stabilizer),
        "negative_control": negative_control(stabilizer, generators),
    }

    assert result["quadratic_evaluation_rank"] == 6
    assert result["required_secant_indices"] == {1: 90, 2: 15, 3: 10}
    assert result["conic_secant_indices"] == {0: 12}
    assert syndrome["distance_counts"] == {0: 1, 1: 60, 2: 1150, 3: 120}
    assert syndrome["distance_two_leader_histogram"] == {1: 900, 2: 150, 3: 100}
    assert chords["edges"] == 30
    assert chords["degree_histogram"] == {5: 12}
    assert all(row["edges"] == 5 for row in chords["color_classes"])
    assert group["stabilizer_size"] == 60
    assert group["element_orders"] == {1: 1, 2: 15, 3: 20, 5: 24}
    assert group["determinant_legendre"] == {1: 60}
    assert result["action_transitivity"] == {"arc_orbit": 6, "conic_orbit": 12, "edge_orbit": 30}
    assert result["coordinate_invariance"] is True
    assert result["negative_control"]["mutated_generator_preserves_arc"] is False
    print(json.dumps(result, sort_keys=True, separators=(",", ":")))


if __name__ == "__main__":
    main()
